core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/MarecoAllowance.java

```diff
@@ -0,0 +1,294 @@
+                    currentAcceleratingSlope.endPosition = previousPosition;
+                    currentAcceleratingSlope.targetSpeed = speed;
+                    meanAcceleration /= accelerationLength;
+                    currentAcceleratingSlope.acceleration = meanAcceleration;
+                    res.add(currentAcceleratingSlope);
+                    // reset meanAcceleration, accelerationLength, and accelerating slope
+                    meanAcceleration = 0;
+                    accelerationLength = 0;
+                    currentAcceleratingSlope =
+                            new AcceleratingSlope(
+                                    0, 0, 0, 0, 0
+                            );
+                }
+                previousAcceleration = naturalAcceleration;
+                previousPosition = position;
+                cursor.nextStep();
+                position = cursor.getPosition();
+                weightForce = TrainPhysicsIntegrator.getWeightForce(rollingStock, path, position);
+                naturalAcceleration = TrainPhysicsIntegrator.computeAcceleration(rollingStock,
+                        rollingResistance, weightForce, speed, 0, 0, 1);
+            }
+            // if the end of the plateau is an accelerating slope
+            if (accelerationLength > 0) {
+                currentAcceleratingSlope.endPosition = previousPosition;
+                currentAcceleratingSlope.targetSpeed = speed;
+                meanAcceleration /= accelerationLength;
+                currentAcceleratingSlope.acceleration = meanAcceleration;
+                res.add(currentAcceleratingSlope);
+                // reset meanAcceleration, accelerationLength, and accelerating slope
+                meanAcceleration = 0;
+                accelerationLength = 0;
+                currentAcceleratingSlope =
+                        new AcceleratingSlope(0, 0, 0, 0, 0);
+            }
+        }
+        return res;
+    }
+
+    /** Generate coasting overlay at a specific position */
+    public Envelope addCoastingCurvesAtPosition(
+            Envelope envelope,
+            List<Double> endOfCoastingPositions
+    ) {
+        for (var position : endOfCoastingPositions) {
+            var builder = OverlayEnvelopeBuilder.backward(envelope);
+            builder.cursor.findPosition(position);
+            var speed = envelope.interpolateSpeed(position);
+            var partBuilder = builder.startContinuousOverlay(COASTING);
+            EnvelopeCoasting.coast(rollingStock, path, timeStep, position, speed, partBuilder, -1);
+            builder.addPart(partBuilder);
+            envelope = builder.build();
+        }
+        return envelope;
+    }
+
+    private Set<EnvelopePart> findBrakingEnvelopeParts(Envelope envelope) {
+        var res = new HashSet<EnvelopePart>();
+        for (var part : envelope) {
+            if (part.meta instanceof DecelerationMeta || part.meta instanceof StopMeta)
```

Comment:
Small nitpick, but it could make sense to have `StopMeta` inherit from `DecelerationMeta`, which would also simplify this condition

Reply:
The whole situation with the typing and content of envelope part meta needs resolving IMO
I've opened #485